### PR TITLE
Implement protonet|| create unit tests

### PR DESCRIPTION
**Pull request by Koushik-S**
This is for developing protonet, which is  an FCN used for creating the masks used in the prediction

### CODE REVIEW

File: model/protonet.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import torch.nn.functional as F
+import numpy as np
+import torch
+from torch import nn
+
+class InterpolateModule(nn.Module):
```
Comment (haraldger) [inline model/protonet.py:6]: This is duplicate

File: model/protonet.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+		self.args = args
+		self.kwdargs = kwdargs
+
+	def forward(self, x):
+		return F.interpolate(x, *self.args, **self.kwdargs)
+
+class InterpolateModule(nn.Module):
+	"""
+	This is a module version of F.interpolate (rip nn.Upsampling).
```
Comment (haraldger) [inline model/protonet.py:23]: Remove these comments

File: model/protonet.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	Any arguments you give it just get passed along for the ride.
+	"""
+
+	def __init__(self, *args, **kwdargs):
+		super().__init__()
+
+		self.args = args
+		self.kwdargs = kwdargs
```
Comment (haraldger) [inline model/protonet.py:31]: We don't need these args and kwdargs

File: model/protonet.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+	def forward(self, x):
+		return F.interpolate(x, *self.args, **self.kwdargs)
```
Comment (haraldger) [inline model/protonet.py:34]: It's better to remove this class and just call F.interpolate directly in the other layers/modules

File: model/protonet.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	
+
+def make_net(in_channels, conf, include_last_relu=True):
```
Comment (haraldger) [inline model/protonet.py:37]: Remove all "conf" references, this is going to be hard for us to create and manage. Hard code layers instead like in the paper

File: model/protonet.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        
+        # Possible patterns:
+        # ( 256, 3, {}) -> conv
+        # ( 256,-2, {}) -> deconv
+        # (None,-2, {}) -> bilinear interpolate
+        # ('cat',[],{}) -> concat the subnetworks in the list
+        #
+        # You know it would have probably been simpler just to adopt a 'c' 'd' 'u' naming scheme.
+        # Whatever, it's too late now.
+        if isinstance(layer_cfg[0], str):
```
Comment (haraldger) [inline model/protonet.py:53]: Instead of these conditional statements, just hard code for our use case. We don't need config files and special cases

File: model/protonet.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        # Don't return a ReLU layer if we're doing an upsample. This probably doesn't affect anything
+        # output-wise, but there's no need to go through a ReLU here.
+        # Commented out for backwards compatibility with previous models
+        # if num_channels is None:
+        #     return [layer]
+        # else:
+        return [layer, nn.ReLU(inplace=True)]
+
+    # Use sum to concat together all the component layer lists
+    net = sum([make_layer(x) for x in conf], [])
```
Comment (haraldger) [inline model/protonet.py:83]: We probably don't need a make_layer function at all. You can just make the layers explicitly.

File: model/protonet.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    if not include_last_relu:
+        net = net[:-1]
+
+    return nn.Sequential(*(net)), in_channels
+
+
+#testing
+def proto_test_shape():
+  conf=[(256, 3, {'padding': 1})] * 3 + [(None, -2, {}), (256, 3, {'padding': 1})] + [(32, 1, {})]
```
Comment (haraldger) [inline model/protonet.py:92]: After you remake the class to not have config objects you can remake the test to.

**Review by haraldger — CHANGES_REQUESTED** [review]
Redesign the class to not be generalized with config files like in the paper's repo, but rather just simplify it and hard code the layers.

File: model/protonet.py (hunk 3 above)
Comment (Koushik-S) [reply model/protonet.py:31]: fuck you

Comment (haraldger) [reply model/protonet.py:31]: D:<
